src/rpc/core_rpc_server.cpp

```diff
@@ -61,34 +63,235 @@
+      std::string serialize(std::string&& res) {
+        std::ostringstream o;
+        epee::serialization::dump_as_json(o, std::move(res), 0 /*indent*/, false /*newlines*/);
+        return o.str();
+      }
+
+      template <typename R = typename RPC::response, std::enable_if_t<!std::is_same<R, std::string>::value, int> = 0>
+      std::string serialize(typename RPC::response&& res) {
+        std::string response;
+        epee::serialization::store_t_to_json(res, response, 0 /*indent*/, false /*newlines*/);
+        return response;
+      }
+    };
+
+    // binary command specialization
+    template <typename RPC>
+    struct reg_helper<RPC, std::enable_if_t<std::is_base_of<BINARY, RPC>::value>> {
+      using Request = typename RPC::request;
+      Request load(rpc_request& request) { 
+        Request req{};
+        if (!request.body.is<string_view>())
+          throw std::runtime_error{"Internal error: can't load binary a RPC command with non-string body"};
+        auto data = request.body.get_unchecked<string_view>();
+        if (!epee::serialization::load_t_from_binary(req, epee::span<const uint8_t>{reinterpret_cast<uint8_t*>(data[0]), data.size()}))
+          throw parse_error{"Failed to parse binary data parameters"};
+        return req;
+      }
+
+      std::string serialize(typename RPC::response&& res) {
+        std::string response;
+        epee::serialization::store_t_to_binary(res, response);
+        return response;
+      }
+    };
+
+    template <typename RPC>
+    void register_rpc_command(std::unordered_map<std::string, std::shared_ptr<const rpc_command>>& regs)
+    {
+      using Request = typename RPC::request;
+      using Response = typename RPC::response;
+      /// check that core_rpc_server.invoke(Request, rpc_context) returns a Response; the code below
+      /// will fail anyway if this isn't satisfied, but that compilation failure might be more cryptic.
+      using invoke_return_type = decltype(std::declval<core_rpc_server>().invoke(std::declval<Request&&>(), rpc_context{}));
+      static_assert(std::is_same<Response, invoke_return_type>::value,
+          "Unable to register RPC command: core_rpc_server::invoke(Request) is not defined or does not return a Response");
+      auto cmd = std::make_shared<rpc_command>();
+      constexpr bool binary = std::is_base_of<BINARY, RPC>::value;
+      cmd->is_public = std::is_base_of<PUBLIC, RPC>::value;
+      cmd->is_binary = binary;
+      cmd->invoke = [](rpc_request&& request, core_rpc_server& server) {
+        reg_helper<RPC> helper;
+        Response res = server.invoke(helper.load(request), std::move(request.context));
+        return helper.serialize(std::move(res));
+      };
+
+      for (const auto& name : RPC::names())
+        regs.emplace(name, cmd);
+    }
+
+    template <typename... RPC>
+    std::unordered_map<std::string, std::shared_ptr<const rpc_command>> register_rpc_commands() {
```

Comment:
I'm not understanding what you found hard to read here; this is a pretty straightforward variadic template that saves needing to repeat a bunch of boilerplate code on every line.

Reply:
It just boils down to, I prefer
```cpp
    std::unordered_map<std::string, std::shared_ptr<const rpc_command>> register_rpc_commands() {
      std::unordered_map<std::string, std::shared_ptr<const rpc_command>> regs;
      register_rpc_command<COMMAND_1>(regs);
      register_rpc_command<COMMAND_2>(regs);
      // ...
      return regs;
    }

    /// Static registration of all rpc commands declared in core_rpc_server_commands_defs.h.
    /// Each of these must also have a core_rpc_server::invoke overload that takes a ::request and
    /// returns a ::response.
    const std::unordered_map<std::string, std::shared_ptr<const rpc_command>> rpc_commands = register_rpc_commands();
    ...
```
Instead of,
```cpp
  namespace {
    // forward declaration; we don't really need to (it is defined just below), but doing it this
    // way keeps the RPC structs close to the top of this file.
    template <typename... T>
    std::unordered_map<std::string, std::shared_ptr<const rpc_command>> register_rpc_commands();
  }

  /// Static registration of all rpc commands declared in core_rpc_server_commands_defs.h.
  /// Each of these must also have a core_rpc_server::invoke overload that takes a ::request and
  /// returns a ::response.
  const std::unordered_map<std::string, std::shared_ptr<const rpc_command>> rpc_commands = register_rpc_commands<
    COMMAND_1,
    COMMAND_2
  >();

    ...
    template <typename... RPC>
    std::unordered_map<std::string, std::shared_ptr<const rpc_command>> register_rpc_commands() {
      std::unordered_map<std::string, std::shared_ptr<const rpc_command>> regs;

#ifdef __cpp_fold_expressions // C++17
      (register_rpc_command<RPC>(regs), ...);
#else
      (void) std::initializer_list<bool>{(register_rpc_command<RPC>(regs), true)...};
#endif

      return regs;
    }
```

And also don't end up reading register_rpc_commands() down the bottom, scan back up to the top to see the variadic args, then go back down and substitute into register_rpc_command to follow the execution, like I did.

Reply:
There are two parts to this, as I understand it.  I've cleaned this up a bit, but I'm not going to avoid templates in a case like this where templates make the code cleaner and DRYer.

1. There *is* a bit of a messiness between the list of types (in `core_rpc_server.cpp`), the type definitions (in `core_rpc_server_command_defs.h`), and the use a forward declaration to keep the type lists higher up in the file.  It's also a bit messy to have the type list only in the function rather than being in a self-contained structure, and a .cpp file is arguably the wrong place to define the available RPC types.

I've addressed these points by moving the list of types into `core_rpc_server_command_defs.h` so that it is in the same place as the definition of those types, which seems cleaner and then avoids the of a forward declaration and specific type list in the .cpp at all.  This also lets any other code that wants to enumerate the list of valid RPC types via the header do so rather than needing to reimplement the list buried in the .cpp file.  This cleans it up quite a bit, but I'm not sure you'll be completely satisfied with it because:

2. You are uncomfortable with templates.

I can't do anything about this except encourage you to get a better grasp of them.  Most of what you suggested here boils down to "I am uncomfortable with templates, and so I prefer <code that uses less templates>" but this is not a convincing argument to me at all: templates are a core part of C++, and let you write more DRY code.

> And also don't end up reading register_rpc_commands() down the bottom, scan back up to the top to see the variadic args, then go back down and substitute into register_rpc_command to follow the execution, like I did.

This argument is just boiling down to not understanding template parameters.  That's *not* how you should be reading code with template parameters.  You wouldn't tell me that I shouldn't use function parameters because it's too hard to scroll between the call point and the function because it *doesn't matter*: the function parameters are *parameters*.  You care that you pass in something valid, and you care that the function operates in a sane way for any possible valid parameter.  You might even trace through with a sample parameter value to help follow the code, but at no point would you ever say "I need to write out all possible inputs and substitute them into this function to understand what it does."

Template parameters are the exact same thing: you don't have to know what they are *at all* to read the code.  Sure, you want to pass in something sane, and sure you want to do the right thing, but this:
```C++
(register_rpc_command<RPC>(regs), ...);
```
is pretty simple code that just means call this function for each RPC argument.  The argument here happens to be a type rather than a value, so you pass it in `<>` instead of inside `()` but I don't have to know what each RPC argument and expand everything to read the code.  (The pre-C++17 `ifdef` is admittedly ugly, but it's there deliberately to show the intention of the code with the pre-C++17 being a workaround to implement the intended code.  This will also go away very soon; the next hard fork will already require a C++17 compiler for lokinet, so there's no point in not also requiring the same for loki-core and eliminating backwards compatibility shims like this one).

But more importantly, this template parameterization makes the code more DRY: if I wanted to rename the function or add a second parameter into the register_rpc_command call, I would just change this one line from:
```C++
(register_rpc_command<RPC>(regs), ...);
```
to:
```C++
(register_rpc_command<RPC>(regs, true), ...);
```
instead of needing to change the call on 78 lines.

Reply:
The issue was the declaration was not localised with the implementation and ended up producing more LoC than it saved. You have a point that you do DRY with generics on the function call, but I'd wager the tradeoff on localizing the code over separating it, improves readability.

What I'm saying is there's a human cost that every person who comes across this code will have to go through to understand what it is doing irrespective of how proficient they are with templates. You are compressing information into lines of code that is separated and needs to be decoded and grokked together. It's intent is clear to the author, on the other hand for anyone else trying to understand it is going to need to glue the parts together to validate their mental model of the code.

I mean, naturally when reading through I'm going to be reading top to bottom and there's a bunch of stuff inbetween the declaration and usage that takes your mental model away from the initial declaration.

Having talked about it, putting the variadic function ontop of the variadic declaration is what should have been my initial comment, it's only 3-4 LoC, my gripe was it being separated, which has been fixed now.

Reply:
> my gripe was it being separated, which has been fixed now.

That's fair, and I am happier with the current implementation approach viewing it from "further away."